src/main/java/com/image/gallery/search/service/impl/CachingImageServiceImpl.java

```diff
@@ -0,0 +1,183 @@
+package com.image.gallery.search.service.impl;
+
+import com.image.gallery.search.exseption.GetDataFromUrlException;
+import com.image.gallery.search.exseption.MyIoException;
+import com.image.gallery.search.model.CroppedImage;
+import com.image.gallery.search.model.FullSizeImage;
+import com.image.gallery.search.model.ImageDetails;
+import com.image.gallery.search.model.Tag;
+import com.image.gallery.search.service.CachingImageService;
+import com.image.gallery.search.service.CroppedImageService;
+import com.image.gallery.search.service.FullSizeImageService;
+import com.image.gallery.search.service.ImageDetailsService;
+import com.image.gallery.search.service.TagService;
+import com.image.gallery.search.util.TokenUtil;
+import java.awt.image.BufferedImage;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import javax.imageio.ImageIO;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpHeaders;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CachingImageServiceImpl implements CachingImageService {
+    private static final String IMAGE_URL = "http://interview.agileengine.com/images";
+    private final ImageDetailsService imageDetailsService;
+    private final FullSizeImageService fullSizeImageService;
+    private final CroppedImageService croppedImageService;
+    private final TagService tagService;
+    private final TokenUtil getTokenUtil;
```

Comment:
```suggestion
    private final TokenUtil tokenUtil;
```
for fields usually, nouns are being used